creusot/src/backend/clone_map.rs

```diff
@@ -373,109 +414,113 @@ impl<'tcx> CloneMap<'tcx> {
-                );
+            if key != self.self_id {
+                self.add_graph_edge(self.self_id, key);
             }
 
             if self.names[&key].kind == Kind::Hidden {
                 continue;
             }
 
-            if traits::still_specializable(self.tcx, param_env, key.did, key.subst) {
-                self.names[&key].opaque();
-            }
+            if let Some((did, subst)) = key.did() {
+                if traits::still_specializable(self.tcx, param_env, did, subst) {
+                    self.names[&key].opaque();
+                }
 
-            if !ctx.is_transparent_from(key.did, self.self_id) {
-                self.names[&key].opaque();
-            }
+                if self.self_did().is_some_and(|self_did| !ctx.is_transparent_from(did, self_did)) {
+                    self.names[&key].opaque();
+                }
 
-            ctx.translate(key.did);
+                ctx.translate(did);
+
+                if util::is_inv_internal(self.tcx, did) && self.clone_level == CloneLevel::Body {
+                    self.clone_tyinv(ctx, subst.type_at(0));
+                }
+
+                self.clone_laws(ctx, did, subst);
+            }
 
-            trace!(
-                "{:?} {:?} has {:?} dependencies",
-                self.names[&key].kind,
-                key,
-                ctx.dependencies(key.did).map(|d| d.len())
-            );
-            self.clone_laws(ctx, key);
             self.clone_dependencies(ctx, key);
         }
     }
 
-    fn clone_dependencies(&mut self, ctx: &mut Why3Generator<'tcx>, key: Key<'tcx>) {
-        let key_dep = DepNode::new(ctx.tcx, key);
-        // Check the substitution for dependencies on closures
-        for ty in key.subst.types().flat_map(|t| t.walk()) {
-            let ty = match ty.unpack() {
-                GenericArgKind::Type(ty) => ty,
-                _ => continue,
-            };
+    fn clone_tyinv(&mut self, ctx: &mut Why3Generator<'tcx>, ty: Ty<'tcx>) {
+        if ty.is_box() || matches!(ty.kind(), TyKind::Param(_)) {
+            return;
+        }
 
-            match ty.kind() {
-                TyKind::Closure(id, subst) => {
-                    let key = (*id, *subst).into();
-                    self.insert(key);
-                    // Sketchy... shouldn't we need to do something to subst?
-                    // eprintln!("{:?}", DepNode::new(ctx.tcx, key));
-                    self.add_graph_edge(key_dep, DepNode::new(ctx.tcx, key));
-                }
-                TyKind::Adt(def, subst) => {
-                    let key = (def.did(), *subst).into();
-                    self.insert(key);
-                    self.add_graph_edge(key_dep, DepNode::new(ctx.tcx, key));
-                }
-                _ => {}
-            }
+        let inv_kind = TyInvKind::from_ty(ty);
+        if let CloneNode::TyInv(self_ty) = self.self_id && TyInvKind::from_ty(self_ty) == inv_kind {
```

Comment:
seems like we could add a `CloneNode::ty_inv_kind()` function which returns `Option<TyInvKind>`.